### PR TITLE
Generate reprepro configuration

### LINKED ISSUE #12: Generate reprepro configuration on config change
When all required charm configurations are set, the reprepro config should be generated.

Spec: https://docs.google.com/document/d/1Gzb5iaWzPvZ0M2HtqFH3ksTsW1arb57R4mM6Pic4y60/edit#

### PR DESCRIPTION
**Pull request by albertodonato**
This generates reprepro config files when the charm configuration has all the needed info.

It also adds a simple config file which is sourced by the mirroring script, to configure the name of the exported mirror.

Fixes #12

### CODE REVIEW

**Comment by albertodonato** [comment]
[Configure reprepro from charm configuration](https://trello.com/c/gYMhvs6n/222-configure-reprepro-from-charm-configuration)

File: unit_tests/test_reprepro.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+
+class ConfigureRepreproTest(TestWithFixtures):
+
+    def setUp(self):
+        super().setUp()
+        self.tempdir = self.useFixture(TempDir())
+
+    @mock.patch('charmhelpers.core.hookenv.charm_dir')
```
Comment (freeekanayaka) [inline unit_tests/test_reprepro.py]: FWIW, you would not need this mocking if you used charm-test.

Comment (albertodonato) [inline unit_tests/test_reprepro.py]: Done. I was using CharmTest in other tests already

File: config.yaml [hunk 2]
```diff
@@ -5,8 +5,16 @@ options:
...
+    description: |
+      The URI of the repository to mirror.
+      The format is the following:
+
+      http[s]://[<user>:<pass>@]<hostname>/<path> <suite> <components>
     default: ""
+  mirror-archs:
+    type: string
+    description: "A comma-separated list of architectures to mirror."
+    default: "sources i386 amd64"
```
Comment (sparkiegeek) [inline config.yaml]: Hmm, this is neither comma separated, not a list of architectures

File: lib/charms/archive_auth_mirror/gpg.py [hunk 3]
```diff
@@ -13,6 +13,6 @@ def import_gpg_keys(mirror_key, sign_key, gnupghome=None):
     imported_sign_key = gpg.import_keys(sign_key)
 
     def fingerprint(imported):
-        return imported.results[0]['fingerprint']
+        return imported.results[0]['fingerprint'][-8:]
```
Comment (sparkiegeek) [inline lib/charms/archive_auth_mirror/gpg.py]: 8 is a bit magical, can we get a comment explaining its significance please?

File: lib/charms/archive_auth_mirror/reprepro.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import getpass
+
+from charmhelpers.core.templating import render
+
+from .utils import get_paths
+
+
+def configure_reprepro(mirror_uri, mirror_archs, mirror_key_fingerprint,
+                       sign_key_fingerprint, get_paths=get_paths):
+    '''Create reprepro configuration files.'''
```
Comment (sparkiegeek) [inline lib/charms/archive_auth_mirror/reprepro.py]: Huh? Triple single quotes are a no-no

as per https://www.python.org/dev/peps/pep-0008/#string-quotes and https://www.python.org/dev/peps/pep-0257/

File: config.yaml (hunk 2 above)
Comment (albertodonato) [inline config.yaml]: Fixed the wording. It's a list of architectures in terms of debian repository.

File: lib/charms/archive_auth_mirror/gpg.py (hunk 3 above)
Comment (albertodonato) [inline lib/charms/archive_auth_mirror/gpg.py]: Added

File: lib/charms/archive_auth_mirror/reprepro.py (hunk 4 above)
Comment (albertodonato) [inline lib/charms/archive_auth_mirror/reprepro.py]: Replaced

File: lib/charms/archive_auth_mirror/utils.py [hunk 5]
```diff
@@ -6,50 +6,53 @@
...
 
 
 def get_website_relation_config(domain):
-    '''Return the configuration for the 'static-website' relation.'''
+    """Return the configuration for the 'static-website' relation."""
     port = 80
     paths = get_paths()
     vhost_config = textwrap.dedent(
-        '''
+        """
```
Comment (sparkiegeek) [inline lib/charms/archive_auth_mirror/utils.py]: :man_shrugging: these could have stayed as '''

File: unit_tests/test_gpg.py [hunk 6]
```diff
@@ -1,13 +1,10 @@
...
-    TempDir,
-)
+from charmtest import CharmTest
 
 from charms.archive_auth_mirror.gpg import import_gpg_keys
 
 
 PUBLIC_KEY_FINGERPRINT = 'E275C4776E00A6BAED081A97DE2922ADDC4EDFF2'
-PUBLIC_KEY_MATERIAL = '''
+PUBLIC_KEY_MATERIAL = """
```
Comment (sparkiegeek) [inline unit_tests/test_gpg.py]: :man_shrugging: these too

File: lib/charms/archive_auth_mirror/reprepro.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                       sign_key_fingerprint, get_paths=get_paths):
+    """Create reprepro configuration files."""
+    paths = get_paths()
+    context = split_repository_uri(mirror_uri)
+    context.update(
+        {'archs': mirror_archs,
+         'mirror_key': mirror_key_fingerprint,
+         'sign_key': sign_key_fingerprint})
+    # explicitly pass owner and group for tests, otherwise root would be used
+    owner = group = getpass.getuser()
```
Comment (sparkiegeek) [inline lib/charms/archive_auth_mirror/reprepro.py]: username and groupname being the same feels close to coincidental

Comment (albertodonato) [inline lib/charms/archive_auth_mirror/reprepro.py]: yeah it's not ideal but it's normally the case (and in real use the code runs as root/root)

**Pull request merged**